### PR TITLE
Lighting system update

### PR DESCRIPTION
**Pull request by Markolie**
**What does this PR do:**
This pull request updates our lighting system so it's entirely up-to-date with /tg/'s and moves lighting to its own subsystem (based on https://github.com/ParadiseSS13/Paradise/pull/11345). Fox mentioned the lag that happens later in the round is caused by lighting, so I'm hopeful this may alleviate some of it.

This was made possible because of https://github.com/ParadiseSS13/Paradise/pull/11343, which ports over the plane_master system. As such, this PR shouldn't be merged until ambient occlusion is in.

I'm not an expert on the lighting system. I broadly know how it works, but the underlying matrices and calculations I'm not familiar with. As such, I'm blindly relying on /tg/'s implementation being superior. Ideally we should test merge this to see to what extent this improves performance.

https://github.com/tgstation/tgstation/pull/25931 has also been ported. This means that darkness no longer uses see_invisible, but lighting_alpha instead. There are now three levels of night vision: certain creatures allow toggling between all levels. Thermals now give some night vision. Mesons no longer fully light up non-visible tiles: these use regular darkness instead. 

This PR also updates starlight so it uses /area/space/nearspace for performance boosts, makes it possible to create rotatium (LSD + teslium + meth) and adds rotatium to vent clogs. It also updates our fullscreen handling so it scales with window size. Lastly, it adds fake ambient occlusion (see https://github.com/ParadiseSS13/Paradise/pull/11343).

**After merging, the following SQL query needs to be executed to enable the ambient occlusion toggle by default:**
```
UPDATE
  player
SET
  toggles = toggles + 2097152
WHERE
  toggles <= 1048576
```

**Changelog:**
:cl: Markolie
add: Ambient occlusion had been added to the game. This represents a shadowy effect on most objects. It can be turned off through the game preferences menu.
tweak: Our lighting system has been updated to /tg/'s latest version. Hopefully this means we'll have less lag later in the round.
tweak: Night vision has been modified. There are now three levels of night vision. Certain creatures that were previously able to toggle night vision can now switch between all levels. Thermals now give some night vision. 
tweak: Mesons now no longer fully light up non-visible turfs. It uses slight darkness instead.
/:cl:

### CODE REVIEW

**Comment by Markolie** [comment]
This PR is pretty bloated right now with commits from the ambient occlusion PR - once that's merged, this should be a bit cleaner.

File: code/modules/lighting/lighting_setup.dm [hunk 1]
```diff
@@ -1,16 +1,12 @@
...
-	for(var/zlevel = 1 to world.maxz)
-		create_lighting_overlays_zlevel(zlevel)
-
-/proc/create_lighting_overlays_zlevel(var/zlevel)
-	ASSERT(zlevel)
-
-	for(var/turf/T in block(locate(1, 1, zlevel), locate(world.maxx, world.maxy, zlevel)))
-		if(!T.dynamic_lighting)
+/proc/create_all_lighting_objects()
+	for(var/area/A in GLOB.all_areas)
```
Comment (Markolie) [inline code/modules/lighting/lighting_setup.dm]: Something in here is causing the lighting subsystem to take 35 seconds to initialize. If anyone has any insight here, I'd love to hear it. This works the same as it does on /tg/, so I'm a bit puzzled.

**Comment by Markolie** [comment]
Oh, and this changes starlight so it uses the /turf/space/nearstation turf to generate starlight like /tg/ does. I haven't updated our maps yet, so starlight isn't present at the moment. My preferred solution would be for this PR to go through first, then I'll make a new PR that makes the map changes. That'll conflict way too hard to include in this PR though.

**Comment by AffectedArc07** [comment]
More of my work cut out for me

File: code/modules/reagents/chemistry/recipes/toxins.dm [hunk 2]
```diff
@@ -174,4 +174,11 @@
...
 	mix_message = "The mixture thins and loses all color."
-	mix_sound = 'sound/goonstation/misc/drinkfizz.ogg'
\ No newline at end of file
+	mix_sound = 'sound/goonstation/misc/drinkfizz.ogg'
+
+/datum/chemical_reaction/rotatium
+	name = "Rotatium"
+	id = "Rotatium"
+	result = "rotatium"
+	required_reagents = list("lsd" = 1, "teslium" = 1, "fentanyl" = 1)
```
Comment (Fox-McCloud) [inline code/modules/reagents/chemistry/recipes/toxins.dm]: we don't have fentanyl; I'd change this to meth, morphine, or hydrocodone, instead.

**Comment by KasparoVy** [comment]
Fuck yeeeesss, lighting PLANE!

File: code/modules/reagents/chemistry/recipes/toxins.dm (hunk 2 above)
Comment (Markolie) [reply code/modules/reagents/chemistry/recipes/toxins.dm]: Ugh, why do we still have reagent strings?

**Comment by Markolie** [comment]
@AffectedArc07 Aye. Hopefully this takes away the final roadblock in your crusade against the process scheduler.

File: code/modules/lighting/lighting_setup.dm (hunk 1 above)
Comment (KasparoVy) [reply code/modules/lighting/lighting_setup.dm]: Shitloads of areas? Maybe we should do `for(var/i in GLOB.all_areas)` then under that do `var/area/a = i` to see if that speeds it up a tiny bit- and for the turfloop below

We've seen on occasion funky shiz like that can yield slight performance improvement

File: code/modules/client/client procs.dm [hunk 3]
```diff
@@ -333,6 +333,8 @@
 	if(!prefs)
 		prefs = new /datum/preferences(src)
 		preferences_datums[ckey] = prefs
+	else
+		prefs.parent = src
```
Comment (Fox-McCloud) [inline code/modules/client/client procs.dm]: What's the need for this?

File: code/modules/client/preference/preferences_mysql.dm [hunk 4]
```diff
@@ -18,7 +18,7 @@
 					ghost_anonsay,
 					exp,
 					clientfps,
-					atklog
+					atklog,
```
Comment (Fox-McCloud) [inline code/modules/client/preference/preferences_mysql.dm]: hmmm

**Comment by AffectedArc07** [comment]
@Markolie really appreciate it.

The process scheduler is going to stop scheduling processes soon.

**Comment by Fox-McCloud** [comment]
After a longggg day of testing, I've pinned down the hangup to a single source: https://github.com/ParadiseSS13/Paradise/blob/d9d975a91f19eb1b6006f18da553688624dfbba0/code/game/machinery/alarm.dm#L378

It doesn't play nice with the changes to area initialization and vent/scrubbers getting added to it. 

i tried to work around it by making scrubbers and vents initialize instead of calling `New`, same with air alarms and anything related to these 3.

For whatever reason it just doesn't like this, and ends up creating 17,000,000 signals at roundstart, which is the reason it delays everything (and no, I'm not exaggerating on that number). 

I'm pretty sure alarm's `receive_signal` can be removed. It's old baycode relating to the atmospherics control computer, which posted signals, at one point. These days it's just a UI that check if an alarm has a mode set, and if it does, just lets you interact with it directly....so there's really no signal handling for alarms anymore, as far as I can tell.

File: code/modules/client/client procs.dm (hunk 3 above)
Comment (Markolie) [reply code/modules/client/client procs.dm]: When you log out, your client is gone which means the preferences loses its client. Now, when you log back in the parent client will be set properly. This has been around for forever, but I guess this is the first time anything needs to call the client.

File: code/modules/client/preference/preferences_mysql.dm (hunk 4 above)
Comment (Markolie) [reply code/modules/client/preference/preferences_mysql.dm]: Will fix that in the AO PR.

Comment (Markolie) [reply code/modules/client/preference/preferences_mysql.dm]: On that note, I need to pay better attention when reverting stuff. That could’ve caused serious issues.

Comment (Fox-McCloud) [reply code/modules/client/preference/preferences_mysql.dm]: Go ahead and fix it here, for reasons discussed on discord.

File: code/__DEFINES/preferences.dm [hunk 5]
```diff
@@ -29,10 +29,12 @@
 #define UI_DARKMODE 131072
 #define DISABLE_KARMA 262144
 #define CHAT_NO_MENTORTICKETLOGS 524288
-
 #define TYPING_ONCE 1048576
+#define AMBIENT_OCCLUSION 2097152
+
+#define TOGGLES_TOTAL 4194304 // If you add or remove a preference toggle above, make sure you update this define with the total value of the toggles combined.
```
Comment (Fox-McCloud) [inline code/__DEFINES/preferences.dm]: `4194303`

Comment (AffectedArc07) [reply code/__DEFINES/preferences.dm]: Didnt I comment this before?

Comment (Markolie) [reply code/__DEFINES/preferences.dm]: Yea, that was on the other PR.

File: code/controllers/subsystem/lighting.dm [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+	for(i in 1 to GLOB.lighting_update_lights.len)
+		var/datum/light_source/L = GLOB.lighting_update_lights[i]
+
+		L.update_corners()
+
+		L.needs_update = LIGHTING_NO_UPDATE
+
+		if(init_tick_checks)
+			CHECK_TICK
+		else if (MC_TICK_CHECK)
```
Comment (Aurorablade) [inline code/controllers/subsystem/lighting.dm]: okay personal nitpick but i always hated the space between the if and the paren thing.

Comment (KasparoVy) [reply code/controllers/subsystem/lighting.dm]: Our style guide agrees haha. No spesses

**Comment by Couls** [comment]
does this mean using mesons will have the others rooms show up dark like in TG?

**Comment by Fox-McCloud** [comment]
Don't forget to update Diona and Shadowling light calculations.

**Comment by Markolie** [comment]
> does this mean using mesons will have the others rooms show up dark like in TG?

That's correct.

File: code/controllers/subsystem/lighting.dm (hunk 6 above)
Comment (Markolie) [reply code/controllers/subsystem/lighting.dm]: I'll have to adjust this. /tg/ doesn't really care about this, which is why it's inconsistent. 

Honestly, I think we shouldn't worry so much about these small issues, like /tg/ does. We don't have anywhere near as many coders or reviewers as them, yet we tend to have much more in-depth reviews and care much more about issues like these. Personally, I think we'd be much better off loosening up in that regard so we can get more (new) stuff in.

**Comment by Markolie** [comment]
Okay, this seems to work fine now. Only things that need doing is (1) starlight space turfs, which will happen in a PR once this is merged and (2) figuring our species see_in_dark, which doesn't really have as much of an impact anymore.

**Comment by KasparoVy** [comment]
Can you define what you mean by 'doesn't have as much of an impact'? Less than it already does? Yikes...

https://github.com/PolarisSS13/Polaris/pull/4626 would be FANTASTIC if it's compatible with our implementation of plane-based lighting.
There was some other PR based a bit more closely on MSO's seeing in dark settings but I lost track of it.

**Comment by Markolie** [comment]
It quite literally does nothing, except if you're wearing mesons which effectively turns those into shitty night vision. Mesons + high species see_in_dark might actually be a bit too strong, come to think of it.

I think the eyes adjusting in the dark may be a little too strong as well, as it means you basically get delayed night vision on certain species.

**Comment by KasparoVy** [comment]
It's live calculation based on brightness of area and parameters/how bright it gets can be adjusted as desired

**Comment by Fox-McCloud** [comment]
Our options are fairly limited, basically. We can do a number of things:

Implement exactly what TG has. This means darksight basically does little, by itself, unless you're wearing mesons, where in it functions exactly like it does on live. This is probably the closest to keeping the system the same.

Implement what TG has, but do something funky with the "nightvision" overlay. Don't let the name trick you; this is just me referencing what it's called in the code. The "nightvision" overlay is basically a tiny little circle around your mob that lets you see a few tiles around you. Hypothetically we could alter this, on a per species basis, or per darksight basis, so that the circle's size depends on your darkvision, but...as this is a fixed screen overlay? Kinda ouchy.

Kill of species darksight entirely. Zoink. Nothing. No one has darksight anymore. Probably not a good idea, I kinda consider this to be giving up.

Give species night eyes. Night eyes are basically an eye type that has full dark vision, but also lets you switch between which lighting planes you want to see. This is really doggone cool, but it may also be a bit strong for the species that have darksight...it also really is only applicable for species who have dark vision range of 8--it wouldn't be all that sensible for those species in between. Taj and Vulp do have a massive weakness to anything light related already, so *maybe* the change could be justified, buttttt I'm not so sure about that, either.

@KasparoVy

**Comment by KasparoVy** [comment]
Thanks for the explanations both of you, I appreciate it
I checked out the PR to a local branch and yeah with 8 darksight you can right-click and examine stuff on tiles for the whole screen but the darkness overlay is quite prohibitive.

Something to fiddle with I guess. I have a couple ideas and a hunch but the only problem I can see is maintaining that base one-tile circle of vision before it expands out to greater vision rather than starting at utter blackness.

**Comment by taukausanake** [comment]
I vote you do the funky /tg/ edits to darksight or kill it completely. You either do it correctly or kill it and hope someone can do it better in the future. Darksight has been a pain of an argument in the past but I really want to see it be good

**Comment by KasparoVy** [comment]
Fair enough. I agree with the TG edits too and will look into whipping up some overlays for each increment of darksight clear up to max and a test tonight if time permits. 

Hopefully it's not too rough making the overlays.

**Comment by Markolie** [comment]
I'm happy with the state of the PR; everything seems to work now. We can figure out the darksight solution in a future PR, as that'll be a little controversial.

**Comment by KasparoVy** [comment]
Pointing at people has the arrows appearing under folks too from what it looks like, not sure if one of the above commits sorted it or not but figured I'd drop a mention.

Plays great!

**Comment by Markolie** [comment]
Thanks for the report @KasparoVy! Looks like the mob overlay plane fix resolved that issue as well.

**Comment by MrMagolor** [comment]
While I could probably just look at tg's wiki for this: what is rotatium? Is it like goon's glowing fliptonium; a chemical that causes anything it is applied to to spin?

**Comment by Markolie** [comment]
It causes your screen to sway left and right.

File: code/_onclick/hud/fullscreen.dm [hunk 7]
```diff
@@ -45,17 +43,38 @@
 		clear_fullscreen(category)
 
 /datum/hud/proc/reload_fullscreen()
-	var/list/screens = mymob.screens
-	for(var/category in screens)
-		mymob.client.screen |= screens[category]
+	if(mymob.client)
+		var/obj/screen/fullscreen/screen
+		var/list/screens = mymob.screens
```
Comment (Fox-McCloud) [inline code/_onclick/hud/fullscreen.dm]: not saying you have to change it, but interesting they don't use `mymob`.

File: code/datums/action.dm [hunk 8]
```diff
@@ -137,9 +137,10 @@
 			var/obj/item/I = target
 			var/old_layer = I.layer
 			var/old_plane = I.plane
-			I.layer = HUD_LAYER_SCREEN + 1
-			I.plane = HUD_PLANE
-			current_button.overlays += I
+			I.layer = FLOAT_LAYER //AAAH
+			I.plane = FLOAT_PLANE //^ what that guy said
```
Comment (Fox-McCloud) [inline code/datums/action.dm]: lol.

File: code/game/atoms.dm [hunk 9]
```diff
@@ -144,9 +145,12 @@
 		alternate_appearances = null
 
 	QDEL_NULL(reagents)
-	invisibility = 101
+	invisibility = INVISIBILITY_MAXIMUM
 	LAZYCLEARLIST(overlays)
 	LAZYCLEARLIST(priority_overlays)
+
+	QDEL_NULL(light)
```
Comment (Fox-McCloud) [inline code/game/atoms.dm]: I'm kinda surprised we never did this before. granted. I believe everything had `set_light(0)` somewhere, but this is still odd this wasn't a thing.

File: code/game/atoms.dm [hunk 10]
```diff
@@ -681,6 +685,7 @@ var/list/blood_splatter_icons = list()
 
 //the sight changes to give to the mob whose perspective is set to that atom (e.g. A mob with nightvision loses its nightvision while looking through a normal camera)
 /atom/proc/update_remote_sight(mob/living/user)
+	user.sync_lighting_plane_alpha()
```
Comment (Fox-McCloud) [inline code/game/atoms.dm]: Why do we have to do this, out of curiosity? TG doesn't.

File: code/game/machinery/alarm.dm [hunk 11]
```diff
@@ -375,38 +375,6 @@
 		if(ATMOS_ALARM_DANGER)
 			icon_state = "alarm1"
 
-/obj/machinery/alarm/receive_signal(datum/signal/signal)
```
Comment (Fox-McCloud) [inline code/game/machinery/alarm.dm]: For anyone wondering, this was used, back in the day for the atmospherics console.

It doesn't use signals anymore to register though so this is literally worthless now--it was also hanging all of game startup.

File: code/game/objects/effects/misc.dm [hunk 12]
```diff
@@ -92,4 +92,36 @@
...
+	if(_duration)
+		QDEL_IN(src, _duration)
+
+/obj/effect/dummy/lighting_obj/moblight
+	name = "mob lighting fx"
+
+/obj/effect/dummy/lighting_obj/moblight/Initialize(mapload, _color, _range, _power, _duration)
+	. = ..()
+	if(!ismob(loc))
+		return INITIALIZE_HINT_QDEL
```
Comment (Fox-McCloud) [inline code/game/objects/effects/misc.dm]: Is this working now? I know you said you had issue with it before?

File: code/_onclick/hud/fullscreen.dm (hunk 7 above)
Comment (Markolie) [reply code/_onclick/hud/fullscreen.dm]: Aye, they use the viewmob in order to support their system where observing someone through OOC lets you see their HUD. We don't have that yet, so no need for that one.

File: code/game/atoms.dm (hunk 9 above)
Comment (Markolie) [reply code/game/atoms.dm]: Aye, very strange indeed.

File: code/modules/awaymissions/zvis.dm [hunk 13]
```diff
@@ -84,9 +84,14 @@
 
 /obj/effect/portal_sensor/proc/check_light()
```
Comment (Fox-McCloud) [inline code/modules/awaymissions/zvis.dm]: what even is this?

File: code/modules/lighting/lighting_corner.dm [hunk 14]
```diff
@@ -51,75 +45,99 @@
...
 
 /datum/lighting_corner/dummy/New()
 	return
+
+
+/datum/lighting_corner/Destroy(var/force)
+	if(!force)
+		return QDEL_HINT_LETMELIVE
+
+	stack_trace("Attempted qdel of a lighting corner.")
```
Comment (Fox-McCloud) [inline code/modules/lighting/lighting_corner.dm]: Ahhh man, TG's stack trace was ***amazing***.

File: code/game/atoms.dm (hunk 10 above)
Comment (Markolie) [reply code/game/atoms.dm]: That's to make sure update_remote_sight always updates the lighting plane alpha, as they usually change the user's lighting_alpha. This way you don't need to repeat it several times in each update proc, and it's not done in update_sight() due to the return on the check.

File: code/modules/awaymissions/zvis.dm (hunk 13 above)
Comment (Markolie) [reply code/modules/awaymissions/zvis.dm]: I have no idea. I didn't want to fix it initially, but decided to anyway. It's in the hotel away mission, it seems.

File: code/game/objects/effects/misc.dm (hunk 12 above)
Comment (Markolie) [reply code/game/objects/effects/misc.dm]: For some reason this one QDEL's fine, but some things don't properly trigger this. Maybe has to do with references.

**Comment by Fox-McCloud** [comment]
Just a few comments/questions.

Also, you may want to update `/mob/dview`

File: code/_onclick/hud/fullscreen.dm (hunk 7 above)
Comment (Fox-McCloud) [reply code/_onclick/hud/fullscreen.dm]: Right-o.

**Comment by Markolie** [comment]
What does dview need? Just max lighting_alpha?

**Comment by Fox-McCloud** [comment]
Oh, I'd just check it out to see if it changed or not; it may not have.

**Comment by Fox-McCloud** [comment]
Lemme know when you checked out `dview`. Once you have and it seems alright, I'll go ahead and merge this.

**Pull request merged**